File: webstore/webstore-java/src/main/java/com/renan/webstore/dto/mapper/CategoryDTOMapper.java
package com.renan.webstore.dto.mapper;

import com.renan.webstore.dto.CategoryDTO;
import com.renan.webstore.model.Category;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Function;

@Component
public class CategoryDTOMapper implements Function<Category, CategoryDTO> {
    
    @Override
    public CategoryDTO apply(Category category) {
        return new CategoryDTO(
                category.getId(),
                category.getName(),
                null
        );
    }
    
    public List<CategoryDTO> convertCategoriesToDTO(List<Category> categories) {
        if (categories == null || categories.isEmpty()) return null;
        List<CategoryDTO> result = new ArrayList<>();
        for (Category cat : categories) {
            if (cat.getUpCategory() == null) {
                result.add(apply(cat));
            }
            categories.remove(cat);
        }
        for (Category cat : categories) {
            for (CategoryDTO dto: result) {
                if (cat.getUpCategory().getId().equals(dto.getId())){
                    if (dto.getSubCategories() == null) dto.setSubCategories(new ArrayList<>());
                    dto.getSubCategories().add(apply(cat));
                }
            }
        }
        return result;
    }
    
}

File: webstore/webstore-kotlin/src/main/kotlin/com/renan/webstore/dto/mapper/CategoryDTOMapper.kt
package com.renan.webstore.dto.mapper

import com.renan.webstore.dto.CategoryDTO
import com.renan.webstore.model.Category
import org.springframework.stereotype.Component
import java.util.function.Function
import kotlin.collections.ArrayList

@Component
class CategoryDTOMapper : Function<Category, CategoryDTO> {

    override fun apply(cat: Category): CategoryDTO {
        return CategoryDTO(
            id = cat.id!!,
            name = cat.name,
            subCategories = null
        )
    }

    fun convertCategoriesToDTO(categories: List<Category>): List<CategoryDTO> {
        val listDTO = ArrayList<CategoryDTO>()
        categories.forEach { cat ->
            if (cat.upperCategory == null) {
                listDTO.add(
                    CategoryDTO(
                        id = cat.id!!,
                        name = cat.name
                    )
                )
            }
        }
        categories.forEach {cat ->
            if (cat.upperCategory != null) {
                listDTO.forEach {dto ->
                    if (cat.upperCategory!!.id == dto.id) {
                        if (dto.subCategories == null) dto.subCategories = ArrayList()
                        (dto.subCategories as ArrayList).add(CategoryDTO(
                            id = cat.id!!,
                            name = cat.name
                        ))
                    }
                }
            }
        }
        return listDTO
    }

}